Starfish/grid_tools/interpolators.py

```diff
@@ -0,0 +1,175 @@
+import itertools
+from collections import OrderedDict
+import logging
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from Starfish import config
+from .utils import determine_chunk_log
+
+
+class IndexInterpolator:
+    """
+    Object to return fractional distance between grid points of a single grid variable.
+
+    :param parameter_list: list of parameter values
+    :type parameter_list: iterable
+    """
+
+    def __init__(self, parameter_list):
+        if not isinstance(parameter_list, np.ndarray):
+            parameter_list = np.array(parameter_list)
+        self.npars = parameter_list.shape[-1]
+        self.parameter_list = np.unique(parameter_list)
+        self.index_interpolator = interp1d(self.parameter_list, np.arange(len(self.parameter_list)), kind='linear')
+
+    def __call__(self, param):
+        """
+        Evaluate the interpolator at a parameter.
+
+        :param param:
+        :type param: list
+        :raises ValueError: if *value* is out of bounds.
+
+        :returns: ((low_val, high_val), (frac_low, frac_high)), the lower and higher bounding points in the grid
+        and the fractional distance (0 - 1) between them and the value.
+        """
+        if len(param) != self.npars:
+            raise ValueError('Incorrect number of parameters. Expected {} but got {}'.format(self.npars, len(param)))
+        try:
+            index = self.index_interpolator(param)
+        except ValueError:
+            raise ValueError("Requested param {} is out of bounds.".format(param))
+        high = np.ceil(index).astype(int)
+        low = np.floor(index).astype(int)
+        frac_index = index - low
+        return ((self.parameter_list[low], self.parameter_list[high]), ((1 - frac_index), frac_index))
+
+
+class Interpolator:
```

Comment:
I think this implementation looks correct, and the test to make sure interpolation at the gridpoint returns the spectrum is good. I think one additional test I did to make sure the interpolation was working was to plot the bounding spectra in the grid, and then plot the interpolated spectra as I varied the parameters in between the grid points. I think I had this set up in a Jupyter notebook with a widget. It's hard to quantify this test but it does make it obvious if something is going wrong.

Reply:
Right, I still really like that idea for exploring the data, still. I don't have a ton of experience with widgets, but I'm pretty sure we could write something like that and package it as a widget, itself. So instead of saving a notebook in the repo (takes up a lot of space) we could actually write a widget into the API and then somebody in a jupyter notebook coud do exploration like
```python
from Starfish.graph_utils import starview
starview(grid=HDF5Interface())
# or 
starview(emulator='emulator.hdf5')
```
To launch in either interpolator or emulator mode.